### DISCUSSION TITLE
Welcome to Icarus Discussions!

### DISCUSSION THREAD

**Discussion by smiklosovic** [instaclustr/cassandra-sidecar#6] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by SuhailRahman** [comment]
@smiklosovic I wanted to understand how does Icarus take snapshots and store it in the cloud. I could observe that it creates a data directory and manifest directory. I wanted to know what this manifest directory contains and how is it being created?

**Comment by smiklosovic** [comment]
Hi @SuhailRahman ,

Icarus does "nothing". Everything is done via Esop. Esop is CLI tool which is embedded into Icarus and invoked programmatically. If you have Icarus instance running in a cluster of e.g. three nodes, Icarus per node, one of Icaruses acts as a coordinator and it will send respective backup requests to each node (and to itself) and it tracks the progress and so on.

Once a backup request reaches Icarus, it will use Esop stuff (very simply said) and it will act as if you invoked Esop from CLI. Very roughly speaking, based on "entitites" field, BackupOperation (it is a class) will take a snapshot by calling Cassandra node via JMX. It invokes StorageService#takeSnapshot and snaphot is taken for that node. Then it will create manifest, based on what files a snapshot has created, then all SSTables are uploaded and manifest (that json file) is uploaded too. Keep in mind that if there is already the same SSTable remotely, it will just skip it from uploading.

Manifest directory contains JSON files. Each file is called "manifest" and it contains the description of the whole backup, each file, with some metadata. This is necessary to upload because based on that manifest there will be files downloaded upon restore. If you want to restore, manifest file is downloaded first and then each file it contains.

The previous two paragraphs are done per each node.

**Reply by SuhailRahman** [reply]
Thank you @smiklosovic this gave me a clear idea of how this works. I also wanted to know how is this manifest created? or what tool is it using to create this manifest file?

**Reply by smiklosovic** [reply]
That is just serialised Java object. There is a class called `Manifest.java` in Esop and I just take that object and I serialise it via Jackson to a file.

**Comment by 2ban** [comment]
Hi, @smiklosovic, I checked new 2.0.0 release, do I have to have Cassandra structure like /var/lib/cassandra/data/data now? Or is there a way to keep /var/lib/cassandra/data?

And I also noticed that RemoveBackupModule is absent in Icarus

**Reply by smiklosovic** [reply]
@2ban remove backup module was released as part of 2.0.2.

**Comment by smiklosovic** [comment]
Hi @2ban 

If you are going to backup, data dirs with your data are specified by "--data-dir" flag which you can specify mutliple times. For Esop invocation, you would do "java esop.jar ... --data-dir /mnt/cassandra/data1 --data-dir /some/other/path/data2 --data-dir /var/lib/data3" ... you got it, right?

For Icarus, that translates to an array:

"dataDirs": [ "/path1", "path2", "path3" ]

I can take a look at removal module, sure.

**Reply by 2ban** [reply]
Hi, @smiklosovic, thanks, I missed that I need to pass an array to "dataDirs" and not a plain string. Now it works as expected.
Waiting for removal module to be added to Icarus.

**Reply by smiklosovic** [reply]
@2ban as said above, remove module is exposed in Icarus since 2.0.2 included.
